src/data/sign.lean

```diff
@@ -0,0 +1,203 @@
+instance : comm_group_with_zero sign_type :=
+{ zero            := 0,
+  one             := 1,
+  mul             := (*),
+  inv             := id,
+  mul_zero        := λ a, by cases a; refl,
+  zero_mul        := λ a, by cases a; refl,
+  mul_one         := λ a, by cases a; refl,
+  one_mul         := λ a, by cases a; refl,
+  mul_inv_cancel  := λ a ha,  by cases a; trivial,
+  mul_comm        := λ a b,   by casesm* _; refl,
+  mul_assoc       := λ a b c, by casesm* _; refl,
+  exists_pair_ne  := ⟨0, 1,   by rintro ⟨⟩⟩,
+  inv_zero        := rfl }
+
+instance : linear_order sign_type :=
+{ le           := (≤),
+  le_refl      := λ a, by cases a; constructor,
+  le_total     := λ a b, by casesm* _; dec_trivial,
+  le_antisymm  := λ a b ha hb, by casesm* _; refl,
+  le_trans     := λ a b c hab hbc, by casesm* _; constructor,
+  decidable_le := le.decidable_rel }
+
+instance : has_distrib_neg sign_type :=
+{ neg_neg := λ x, by cases x; refl,
+  neg_mul := λ x y, by casesm* _; refl,
+  mul_neg := λ x y, by casesm* _; refl,
+..sign_type.has_neg }
+
+/-- `sign_type` is equivalent to `fin 3`. -/
+def fin3_equiv : sign_type ≃* fin 3 :=
+{ to_fun :=  λ a, a.rec_on 0 (-1) 1,
+  inv_fun := λ a, match a with
+    | ⟨0, h⟩   := 0
+    | ⟨1, h⟩   := 1
+    | ⟨2, h⟩   := -1
+    | ⟨n+3, h⟩ := (h.not_le le_add_self).elim
+  end,
+  left_inv :=  λ a, by cases a; refl,
+  right_inv := λ a, match a with
+    | ⟨0, h⟩   := rfl
+    | ⟨1, h⟩   := rfl
+    | ⟨2, h⟩   := rfl
+    | ⟨n+3, h⟩ := (h.not_le le_add_self).elim
+  end,
+  map_mul' := λ x y, by casesm* _; refl }
+
+section cast
+
+variables {α : Type*} [has_zero α] [has_one α] [has_neg α]
+
+/-- Turn a `sign_type` into zero, one, or minus one. This is a coercion instance, but note it is
+only a `has_coe_t` instance: see note [use has_coe_t]. -/
+def cast : sign_type → α
+| zero :=  0
+| pos  :=  1
+| neg  := -1
+
+instance : has_coe_t sign_type α := ⟨cast⟩
+
+@[simp] lemma cast_eq_coe (a : sign_type) : (cast a : α) = a := rfl
+
+@[simp] lemma coe_zero : ↑(0 : sign_type) = (0 : α) := rfl
+@[simp] lemma coe_one  : ↑(1 : sign_type) = (1 : α) := rfl
+@[simp] lemma coe_neg_one : ↑(-1 : sign_type) = (-1 : α) := rfl
+
+end cast
+
+/-- `sign_type.cast` as a `mul_with_zero_hom`. -/
+@[simps] def cast_hom {α} [mul_zero_one_class α] [has_distrib_neg α] : sign_type →*₀ α :=
+{ to_fun    := cast,
+  map_zero' := rfl,
+  map_one'  := rfl,
+  map_mul'  := λ x y, by cases x; cases y; simp }
+
+end sign_type
+
+variables {α : Type*}
+
+open sign_type
+
+section preorder
+
+variables [has_zero α] [preorder α] [decidable_rel ((<) : α → α → Prop)] {a : α}
+
+/-- The sign of an element is 1 if it's positive, -1 if negative, 0 otherwise. -/
+def sign : α →o sign_type :=
+⟨λ a, if 0 < a then 1 else if a < 0 then -1 else 0, λ a b h, begin
+  dsimp,
+  split_ifs; try {constructor},
+  { cases lt_irrefl 0 (h_1.trans $ h.trans_lt h_3) },
+  { cases h_2 (h_1.trans_le h) },
+  { cases h_2 (h.trans_lt h_4) }
```

Comment:
This is referring to auto-generated names. Can you please rewrite?

Reply:
done, but I think it made it a bit more messy